presto-main/src/main/java/io/prestosql/sql/analyzer/FeaturesConfig.java

```diff
@@ -121,6 +121,7 @@
     private boolean preferPartialAggregation = true;
     private boolean optimizeTopNRowNumber = true;
     private boolean workProcessorPipelines;
+    private boolean removeRedundantSort = true;
```

Comment:
You added getter/setter last, but this isn't the last field. This class is messy.

Reply:
Let's think again about example that you also mentioned in your blog:
```sql
SELECT *, row_number() OVER ()
FROM (SELECT * FROM nation ORDER BY name DESC)
```

the proper way of writing this is:

```sql
SELECT *, row_number() OVER (ORDER BY name DESC)
FROM nation
```

Is the proper way a drop-in replacement?
Before the change, the old way would use distributed sort.
Does the standard-compliant version use distributed sort too?

Reply:
If the proper way is not equally performant yet, we shouldn't have the feature on by default just yet.

Reply:
I believe we added distributed sort specifically for the unpartitioned window function case, so I think that should be the case. I’ll double check

Reply:
> You added getter/setter last, but this isn't the last field. 

I added it right below "work processor pipelines" in both cases. The other ones are out of order.

> This class is messy.

100% agree

Reply:
> Is the proper way a drop-in replacement?
> Before the change, the old way would use distributed sort.

Regardless, a query like this is just broken, semantically speaking. Almost every case I've seen is people relying on ORDER BY for aggregates, since 1) we didn't use to support ordered aggregates until sometime last year, 2) window functions have always supported ORDER BY.

For anyone relying on that behavior, there's the fallback of disabling the config flag (for now).

Reply:
It turns out we currently don't run distributed sorts for unpartitioned window functions with an ORDER BY clause. Here's a PR to add it: https://github.com/prestosql/presto/pull/821

Reply:
> I believe we added distributed sort specifically for the unpartitioned window function case, so I think that should be the case. I’ll double check

Distributed sort utilizes entire cluster memory, while `WindowOperator` does not (it collects partition rows).

Keep in mind that unordered row_number window function was replaced with `RowNumberNode` operator (See: `io.prestosql.sql.planner.optimizations.WindowFilterPushDown.Rewriter#canReplaceWithRowNumber`). This means that with such hacky:
```
SELECT *, row_number() OVER ()
FROM (SELECT * FROM nation ORDER BY name DESC)
```

utilize distributed sort and skip `WindowOperator` completely. However, for:
```
SELECT *, row_number() OVER (ORDER BY name DESC)
FROM nation
```
It will still execute `WindowOperator` which will collect all partition data (thus mitigating benefits of distributed sort).

Maybe we should improve `WindowFilterPushDown` rule too

Reply:
@sopel39 good point. So in https://github.com/prestosql/presto/pull/821 when insert Sort below WindowNode, we could convert the WindowNode to RowNumberNode (in `row_number()` case of course). (I think we should move continue this discussion under #821 though).

Reply:
BTW, for this shape:

```
SELECT *, row_number() OVER ()
FROM (SELECT * FROM nation ORDER BY name DESC)
```

even if the ORDER BY weren't removed, there's no guarantee that the row_number() function (or any other window function, for that matter) will not load all the data in memory and shuffle it in some way before producing its results. So it could break at any time for other reasons unrelated to this optimization, so I'm not too concerned about it, especially since we have an escape hatch.